pkg/controller/atlasdeployment/customzonemapping.go

```diff
@@ -0,0 +1,167 @@
+package atlasdeployment
+
+import (
+	"context"
+	"fmt"
+
+	"go.mongodb.org/atlas/mongodbatlas"
+
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/atlasdeployment/globaldeployment"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+)
+
+func EnsureCustomZoneMapping(service *workflow.Context, groupID string, customZoneMappings []mdbv1.CustomZoneMapping, deploymentName string) workflow.Result {
+	result := syncCustomZoneMapping(context.Background(), service, groupID, deploymentName, customZoneMappings)
+	if !result.IsOk() {
+		service.SetConditionFromResult(status.CustomZoneMappingReadyType, result)
+		return result
+	}
+
+	if customZoneMappings == nil {
+		service.UnsetCondition(status.CustomZoneMappingReadyType)
+		service.EnsureStatusOption(status.AtlasDeploymentCustomZoneMappingOption(nil))
+	} else {
+		service.SetConditionTrue(status.CustomZoneMappingReadyType)
+	}
+
+	return result
+}
+
+func syncCustomZoneMapping(ctx context.Context, service *workflow.Context, groupID string, deploymentName string, customZoneMappings []mdbv1.CustomZoneMapping) workflow.Result {
+	logger := service.Log
+	err := verifyZoneMapping(customZoneMappings)
+	if err != nil {
+		return workflow.Terminate(workflow.CustomZoneMappingReady, err.Error())
+	}
+	_, existingZoneMapping, err := globaldeployment.GetGlobalDeploymentState(ctx, service.Client, groupID, deploymentName)
+	if err != nil {
+		return workflow.Terminate(workflow.CustomZoneMappingReady, fmt.Sprintf("Failed to get zone mapping state: %v", err))
+	}
+	logger.Debugf("Existing zone mapping: %v", existingZoneMapping)
+	var customZoneMapping status.CustomZoneMapping
+	zoneMappingMap, err := getZoneMappingMap(ctx, service.Client, groupID, deploymentName)
+	if err != nil {
+		return workflow.Terminate(workflow.CustomZoneMappingReady, fmt.Sprintf("Failed to get zone mapping map: %v", err))
+	}
+
+	if shouldAdd, shouldDelete := compareZoneMappingStates(existingZoneMapping, customZoneMappings, zoneMappingMap); shouldDelete || shouldAdd {
+		skipAdd := false
+		if shouldDelete {
+			err = deleteZoneMapping(ctx, service.Client, groupID, deploymentName)
+			if err != nil {
+				skipAdd = true
+				logger.Errorf("Failed to sync zone mapping: %v", err)
```

Comment:
Please don't put capital letters to the beginning of errors messages

Reply:
Done